File: src/main/java/io/rsocket/cli/http2/Http2DuplexConnection.java
package io.rsocket.cli.http2;

import io.netty.buffer.ByteBuf;
import io.netty.buffer.ByteBufUtil;
import io.netty.buffer.CompositeByteBuf;
import io.netty.buffer.Unpooled;
import io.netty.channel.ChannelHandlerContext;
import io.rsocket.DuplexConnection;
import io.rsocket.Frame;
import io.rsocket.exceptions.ConnectionCloseException;
import io.rsocket.transport.netty.RSocketLengthCodec;
import java.net.URI;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.logging.Logger;
import org.eclipse.jetty.http.HttpField;
import org.eclipse.jetty.http.HttpFields;
import org.eclipse.jetty.http.HttpURI;
import org.eclipse.jetty.http.HttpVersion;
import org.eclipse.jetty.http.MetaData;
import org.eclipse.jetty.http2.ErrorCode;
import org.eclipse.jetty.http2.api.Session;
import org.eclipse.jetty.http2.api.Stream;
import org.eclipse.jetty.http2.frames.DataFrame;
import org.eclipse.jetty.http2.frames.HeadersFrame;
import org.eclipse.jetty.http2.frames.ResetFrame;
import org.eclipse.jetty.util.Callback;
import org.eclipse.jetty.util.Promise;
import org.reactivestreams.Publisher;
import reactor.core.publisher.DirectProcessor;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import reactor.core.publisher.MonoProcessor;

public class Http2DuplexConnection implements DuplexConnection {
  private static final Logger log = Logger.getLogger(Http2DuplexConnection.class.getName());
  private final MonoProcessor<Void> onClose = MonoProcessor.create();
  private final DirectProcessor<Frame> receive = DirectProcessor.create();
  private final MyRSocketLengthCodec codec = new MyRSocketLengthCodec();

  private CompositeByteBuf existingData = Unpooled.compositeBuffer();

  private Stream.Listener responseListener =
      new Stream.Listener.Adapter() {
        @Override
        public void onData(Stream stream, DataFrame frame, Callback callback) {
          ByteBuffer fd = frame.getData();

          try {
            existingData.addComponent(true, Unpooled.copiedBuffer(fd));
            while (true) {
              ByteBuf sliceData = (ByteBuf) codec.decode(null, existingData);

              if (sliceData != null) {
                receive.onNext(Frame.from(sliceData));
              } else {
                break;
              }
            }
            callback.succeeded();
          } catch (Exception e) {
            callback.failed(e);
          }
        }

        @Override
        public void onHeaders(Stream stream, HeadersFrame frame) {
          log.fine(frame.getMetaData().toString());
          for (HttpField e : frame.getMetaData().getFields()) {
            log.fine(e.getName() + ": " + e.getValue());
          }

          MetaData.Response response = (MetaData.Response) frame.getMetaData();
          if (response.getStatus() == 200) {
            log.info("connected");
          } else {
            receive.onNext(
                Frame.Error.from(
                    0, new ConnectionCloseException("non 200 response: " + response.getStatus())));
            close().subscribe();
          }
        }
      };

  private Stream stream;

  @Override
  public Mono<Void> send(Publisher<Frame> frame) {
    return Flux.from(frame)
        .doOnNext(
            f -> {
              stream.data(
                  dataFrame(f),
                  new Callback() {
                    @Override
                    public void failed(Throwable x) {
                      receive.onError(x);
                      close().subscribe();
                    }
                  });
            })
        .then();
  }

  @Override
  public Flux<Frame> receive() {
    return receive;
  }

  @Override
  public double availability() {
    return stream.isClosed() ? 0.0 : 1.0;
  }

  @Override
  public Mono<Void> close() {
    return Mono.defer(
        () -> {
          // TODO listen to callback
          if (!stream.isClosed()) {
            stream.reset(
                new ResetFrame(stream.getId(), ErrorCode.STREAM_CLOSED_ERROR.code),
                new Callback() {
                  @Override
                  public void failed(Throwable x) {}
                });
          }
          onClose.onComplete();
          return onClose;
        });
  }

  @Override
  public Mono<Void> onClose() {
    return onClose;
  }

  public static Mono<Http2DuplexConnection> create(
      Session session, URI uri, Map<String, String> headers) {
    return Mono.create(
        s -> {
          Http2DuplexConnection c = new Http2DuplexConnection();

          headers.forEach((k, v) -> log.fine(k + ": " + v));

          session.newStream(
              headerFrame(uri, headers),
              new Promise<Stream>() {
                @Override
                public void succeeded(Stream result) {
                  c.stream = result;
                  s.success(c);
                }

                @Override
                public void failed(Throwable x) {
                  s.error(x);
                }
              },
              c.responseListener);
        });
  }

  private DataFrame dataFrame(Frame f) {
    return new DataFrame(stream.getId(), f.content().nioBuffer(), false);
  }

  private static HeadersFrame headerFrame(URI uri, Map<String, String> headers) {
    HttpFields requestFields = new HttpFields();
    for (Map.Entry<String, String> e : headers.entrySet()) {
      requestFields.put(e.getKey(), e.getValue());
    }

    MetaData.Request request =
        new MetaData.Request(
            "POST", new HttpURI(uri.toString()), HttpVersion.HTTP_2, requestFields);
    return new HeadersFrame(request, null, false);
  }

  public static class MyRSocketLengthCodec extends RSocketLengthCodec {
    @Override
    public Object decode(ChannelHandlerContext ctx, ByteBuf in) throws Exception {
      return super.decode(ctx, in);
    }
  }
}

File: src/main/kotlin/io/rsocket/cli/http2/Http2DuplexConnection.kt
package io.rsocket.cli.http2

import io.netty.buffer.ByteBuf
import io.netty.buffer.Unpooled
import io.netty.channel.ChannelHandlerContext
import io.rsocket.DuplexConnection
import io.rsocket.Frame
import io.rsocket.exceptions.ConnectionCloseException
import io.rsocket.transport.netty.RSocketLengthCodec
import org.eclipse.jetty.http.HttpFields
import org.eclipse.jetty.http.HttpURI
import org.eclipse.jetty.http.HttpVersion
import org.eclipse.jetty.http.MetaData
import org.eclipse.jetty.http2.ErrorCode
import org.eclipse.jetty.http2.api.Session
import org.eclipse.jetty.http2.api.Stream
import org.eclipse.jetty.http2.frames.DataFrame
import org.eclipse.jetty.http2.frames.HeadersFrame
import org.eclipse.jetty.http2.frames.ResetFrame
import org.eclipse.jetty.util.Callback
import org.eclipse.jetty.util.Promise
import org.reactivestreams.Publisher
import reactor.core.publisher.DirectProcessor
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import reactor.core.publisher.MonoProcessor
import java.net.URI
import java.util.logging.Logger

class Http2DuplexConnection : DuplexConnection {
  private val onClose = MonoProcessor.create<Void>()
  private val receive = DirectProcessor.create<Frame>()
  private val codec = MyRSocketLengthCodec()

  private val existingData = Unpooled.compositeBuffer()

  private val responseListener = object : Stream.Listener.Adapter() {
    override fun onData(stream: Stream?, frame: DataFrame?, callback: Callback) {
      val fd = frame!!.data

      try {
        existingData.addComponent(true, Unpooled.copiedBuffer(fd))
        while (true) {
          val sliceData = codec.decode(null, existingData)

          if (sliceData != null) {
            receive.onNext(Frame.from(sliceData))
          } else {
            break
          }
        }
        callback.succeeded()
      } catch (e: Exception) {
        callback.failed(e)
      }

    }

    override fun onHeaders(stream: Stream?, frame: HeadersFrame?) {
      log.fine(frame!!.metaData.toString())
      for (e in frame.metaData.fields) {
        log.fine(e.name + ": " + e.value)
      }

      val response = frame.metaData as MetaData.Response
      if (response.status == 200) {
        log.info("connected")
      } else {
        receive.onNext(
            Frame.Error.from(
                0, ConnectionCloseException("non 200 response: " + response.status)))
        close().subscribe()
      }
    }
  }

  private var stream: Stream? = null

  override fun send(frame: Publisher<Frame>): Mono<Void> {
    return Flux.from(frame)
        .doOnNext { f ->
          stream!!.data(
              dataFrame(f),
              object : Callback {
                override fun failed(x: Throwable?) {
                  receive.onError(x!!)
                  close().subscribe()
                }
              })
        }
        .then()
  }

  override fun receive(): Flux<Frame> = receive

  override fun availability(): Double = if (stream!!.isClosed) 0.0 else 1.0

  override fun close(): Mono<Void> = Mono.defer {
    // TODO listen to callback
    if (!stream!!.isClosed) {
      stream!!.reset(
          ResetFrame(stream!!.id, ErrorCode.STREAM_CLOSED_ERROR.code),
          object : Callback {
            override fun failed(x: Throwable?) {}
          })
    }
    onClose.onComplete()
    onClose
  }

  override fun onClose(): Mono<Void> = onClose

  private fun dataFrame(f: Frame): DataFrame =
      DataFrame(stream!!.id, f.content().nioBuffer(), false)

  class MyRSocketLengthCodec : RSocketLengthCodec() {
    @Throws(Exception::class)
    public override fun decode(ctx: ChannelHandlerContext?, `in`: ByteBuf): ByteBuf? {
      return super.decode(ctx, `in`) as ByteBuf?
    }
  }

  companion object {
    private val log = Logger.getLogger(Http2DuplexConnection::class.java.name)

    fun create(
        session: Session, uri: URI, headers: Map<String, String>): Mono<Http2DuplexConnection> {
      return Mono.create { s ->
        val c = Http2DuplexConnection()

        headers.forEach { k, v -> log.fine(k + ": " + v) }

        session.newStream(
            headerFrame(uri, headers),
            object : Promise<Stream> {
              override fun succeeded(result: Stream?) {
                c.stream = result
                s.success(c)
              }

              override fun failed(x: Throwable?) {
                s.error(x!!)
              }
            },
            c.responseListener)
      }
    }

    private fun headerFrame(uri: URI, headers: Map<String, String>): HeadersFrame {
      val requestFields = HttpFields()
      for ((key, value) in headers) {
        requestFields.put(key, value)
      }

      val request = MetaData.Request(
          "POST", HttpURI(uri.toString()), HttpVersion.HTTP_2, requestFields)
      return HeadersFrame(request, null, false)
    }
  }
}
